File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
       if(root==NULL) return 0;
       int maxval = root->val;
       maxPathSumHelper(root,maxval); 
       return maxval;
    }
    int maxPathSumHelper(TreeNode* root,int& maxval){
        if(root==NULL) return 0;
        int lsum = max(0,maxPathSumHelper(root->left,maxval));
        int rsum = max(0,maxPathSumHelper(root->right,maxval));
        maxval = max(maxval,lsum+rsum+root->val);
        return root->val + max(lsum,rsum);
    }
};

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int maxVal ;
    public int maxPathSum(TreeNode root) {
        if(root == null) return 0;
        maxVal = root.val;
        Helper(root);
        return maxVal;
    }
    int Helper(TreeNode root)
    {
        if(root==null) return 0;
        int leftPath = Math.max(Helper(root.left),0);
        int rightPath = Math.max(Helper(root.right),0);
        maxVal = Math.max(leftPath+rightPath+root.val,maxVal);
        return root.val + Math.max(leftPath,rightPath);
    }
}